### ISSUE TITLE
Architecture for 2.0

### ISSUE THREAD

**Issue by davidmehren** [codimd/server#443]
Hi everyone!

@InnayTool and I prepared a proposal for the architecture of the new API backend we have to build for 2.0. For some topics we don't have a concrete proposal yet, these will get their own issues to discuss shortly. Please don't hesitate to comment, we appreciate your feedback!

---

This proposal is mostly based on the "Bulletproof node.js project architecture" [blogpost](https://softwareontheroad.com/ideal-nodejs-project-structure/), with a few modifications. If you want a less bullet-point-ish description, it might make sense to read the blog.

- 3 layers: controller, services, data access
    - a controller recieves HTTP requests, parses & validates the data, creates a data transfer object (DTO) and calls one(!) service
    - a service contains business logic and calls other services or the data access layer. 
    - the data access layer (in our case probably a sequelize model) is responsible for storing and retrieving objects in the database. This is the only place a 'real' model is used (in contrast to DTOs everywhere else).
- We might use an event bus to asynchronously do stuff, e.g. to persist changes notes to the database
    - ideas for concrete patterns wanted!
- Dependencies of services/controllers are provided using dependency injection. This greatly improves testability.
- Things from the blogpost we don't need/will do differently: cron jobs (don't need that[yet?]), configuration (we use yaml and env vars instead of only env vars) 
- folder structure:
```
/src
  index.ts
  /api
    /internal
      user.ts <- UserController
      note.ts <- NoteController
    /stable
      ...
  /user <- directories reflect the API subpaths
    user.interfaces.ts <- interfaces for user DTOs
    user.model.ts <- database model
    user.service.ts <- UserService
    ...
  /note
    note.service.ts
    note-export.service.ts <- more than 1 service possible
    ...
  /realtime
    /ot <- contains old OT code, here be dragons
    realtime.service.ts <- does realtime stuff
    ...
  ...
```
- Files are [bundled by component, not by technical role](https://github.com/goldbergyoni/nodebestpractices/blob/master/sections/projectstructre/breakintcomponents.md)

### Code Style
- [Don't use callbacks](https://github.com/goldbergyoni/nodebestpractices/blob/master/sections/errorhandling/asyncerrorhandling.md), use async/await instead.
- `import` instead of `require`
- We already have plenty more ESLint rules.

### Server
We use Express and Socket.io already and to date no big reason to change that came up. 

### TypeScript Annotations for Express
Discussion in #444

### Dependency Injection
Discussion in #445

### Tests
Discussion in #446

### ORM & Migrations
We already use Sequelize 5 and sequelize-typescript in the 2.x branch. Currently, the latter doesn't support Sequelize 6 but that seems to be [in the works](https://github.com/RobinBuschmann/sequelize-typescript/pull/807). To improve the migration situation, 
Sheogorath started working on using umzug instead of sequelize-cli in [#384](https://github.com/codimd/server/pull/384)

- [sequelize](https://github.com/sequelize/sequelize)
    - Upgrade 5 -> 6
- [sequelize-typescript](https://github.com/RobinBuschmann/sequelize-typescript)
- [umzug](https://github.com/sequelize/umzug)

Discussion about supported databases in #447

### Validation
Discussion in #448

### Logging

As winston seems to perform reasonably well in the current backend and no big problem with it could be found, we'll just keep it.

- [winston](https://github.com/winstonjs/winston)

### Auth Provider
Passport seems to be the way to go here, but the User model needs some refinement.

- [passport](https://github.com/jaredhanson/passport)

### Security

- [helmet](https://helmetjs.github.io/)

Additional ideas by @SISheogorath :
- [express-rate-limit](https://www.npmjs.com/package/express-rate-limit)
  - See also #442
- [jsonschema](https://www.npmjs.com/package/jsonschema) (for validation)
- [safe-regex](https://www.npmjs.com/package/safe-regex) (for all our own regexes)
- [sandbox](https://www.npmjs.com/package/sandbox)/[vm2](https://www.npmjs.com/package/vm2) (could be used for validation, to prevent bad things from happening)
- https://www.npmjs.com/package/eslint-plugin-security

### Configuration

- Validate config using [convict](https://github.com/mozilla/node-convict). PR for this: https://github.com/codimd/server/pull/214

### Errors
Useful reference: https://www.joyent.com/node-js/production/design/errors

- [Handle errors centrally.](https://github.com/goldbergyoni/nodebestpractices/blob/master/sections/errorhandling/centralizedhandling.md)

**Blog posts/websites we found helpful:**

- https://softwareontheroad.com/ideal-nodejs-project-structure/
- https://blog.codecentric.de/2019/05/web-anwendungen-mit-node-express-typescript-entwickeln/
- https://stackoverflow.com/questions/5778245/expressjs-how-to-structure-an-application
- https://github.com/goldbergyoni/nodebestpractices

**Comment by truh** [comment]
Have you checked out [tsoa](https://tsoa-community.github.io/docs/)? It seems pretty in line with the architecture. The OpenAPI compliance could be huge for a potential ecosystem applications and tools around HedgeDoc.

**Comment by davidmehren** [comment]
Thanks for the suggestion, @truh! Actually, the topic of using a framework came up multiple times. Until now, we always decided to not go the framework route, because we would then be completely bound to that specific framework. If the framework suddenly became unmaintained, we would probably have to rewrite a significant part of our code to switch to something else.

Nonetheless, while researching possible libraries for the use-cases mentioned above, @InnayTool and I realized that there are frameworks that provide a significant part or even all of the functionality we would have to piece together from different libs ourselves.

Therefore we looked at the three most promising libraries in more detail: [tsoa](https://tsoa-community.github.io/docs/), [Ts.ED](https://tsed.io/) and [NestJS](https://nestjs.com/). Quickly it became clear that only NestJS meets our requirements for sustainability: It has [multiple corporate sponsors](https://github.com/nestjs/nest#support) and a healthy base of [financial supporters](https://opencollective.com/nest). Therefore we deemed it unlikely for the project to become abandoned. Ts.ED only gets a comparatively [small amount](https://opencollective.com/tsed) of donations, for tsoa we couldn't find any details on funding or sponsors.

Lets have a look at the features. NestJS supports:
- [Annotations for Express](https://docs.nestjs.com/controllers#routing)
- [Dependency Injection](https://docs.nestjs.com/providers#dependency-injection)
- [Testing](https://docs.nestjs.com/fundamentals/testing): Jest and Supertest as default, but compatible with any other framework
- [ORMs](https://docs.nestjs.com/techniques/database): Sequelize and TypeORM
- [Validation](https://docs.nestjs.com/techniques/validation#auto-validation): even in a nicer way than we could build ourselves

It also encourages the layered architecture we want to use.

I have to say, I am quite tempted to go with NestJS, as it combines all features we want with a (as far as we can guesstimate) low risk of becoming abandoned. @codimd/developers what's your opinion on this?

**Comment by SISheogorath** [comment]
To me, as with every framework it raises two important questions:

1. How hurtful is it to use it? Docs are usually looking great, but basically when you get your hands dirty you'll notice whether or not, it's really what one wants to work with.
2. What are the boundaries of the framework (what rules does it enforce) and what if we want/have to go around it, how much does that hurt us?

We will definitely hit a point in time, where we have to do a dirty hack in order to realize something and that's why I'm usually sceptical about frameworks. They are not designed to do that. They are there to keep you away from dirty hacks.

General attitude of the NestJS framework seems to be good, as it exposes all API directory (if the philosophy section of the documentation isn't lying) and this way allows the needed flexibility. Also the fact that in the past years there was simply one major release per year and the amount of breaking changes was moderate, I think it might be worth a try.

**Comment by davidmehren** [comment]
@InnayTool and I played around with NestJS in the last week. Everything worked nicely, but we hit one roadblock: Because Sequelize objects are also used to access the database (ActiveRecord pattern), they contain a bunch of internal properties. One of these properties causes an infinite loop in [class-transformer](https://github.com/typestack/class-transformer), which means that we can't directly send an object from the database to a client, but have to strip the extra properties manually first. Therefore we switched to TypeORM (the other ORM supported by NestJS), which works out-of-the box. We then experimented with accessing an old database (created with Sequelize) with TypeORM, which also worked after a bit of juggling. We therefore are pretty sure that it's possible to write migrations from our old db scheme to a new TypeORM-based one.

NestJS also solves all our library-questions. It actually uses `typestack/routing-controllers`, which was our favorite in #444.

tl;dr: @InnayTool and I are pretty happy with NestJS with TypeORM. As there were no further opinions on this topic, let's go ahead with Nest!
